### PR TITLE
feat(core): improve alias validation

### LINKED ISSUE #2154: Move alias validation from runtime to static
**Chore summary**
Currently if you use an alias in `given` that isn't defined, spectral detects the issue while running `lint` against an openapi file but not while validating the ruleset. 

**Tasks**
- [ ] Validation should fail if an alias is being used that isn't defined

### PR DESCRIPTION
**Pull request by P0lip**
Closes #2154 

**Checklist**

- [x] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

**Changes**

- [extracted](https://github.com/stoplightio/spectral/pull/2164/files#diff-2ac12b44fc3e0b0c43b42e64947c022ccc9543625ad3b6a6efcba574219f8068L154) `resolveAlias` method out of the `Rule` class and moved it to a [separate file](https://github.com/stoplightio/spectral/blob/209fdd558d687c4c82114a1fd68bbba6287f84ab/packages/core/src/ruleset/alias.ts) so it can be used elsewhere
  - the function itself remained pretty much unchanged
 
- `x-spectral-runtime` keyword takes new option, namely `alias` -> https://github.com/stoplightio/spectral/blob/209fdd558d687c4c82114a1fd68bbba6287f84ab/packages/core/src/ruleset/validation/ajv.ts#L62
  - a side change was the adaption of the same validation approach for functions, so that it uses a similar mechanism, [here](https://github.com/stoplightio/spectral/pull/2164/files#diff-29b1db27ac801e0dec894b898b15a315e649cfd54e5280cc45aaf5a800ce63bfR52-R61) 
  
- adjusted [packages/core/src/ruleset/meta/shared.json](https://github.com/stoplightio/spectral/blob/209fdd558d687c4c82114a1fd68bbba6287f84ab/packages/core/src/ruleset/meta/shared.json) to make use of new option ^
  - almost all changes in tests are related to the schema changes I've made

### CODE REVIEW

File: docs/guides/4-custom-rulesets.md [hunk 1]
```diff
@@ -396,6 +396,48 @@ aliases:
...
+        then:
+          field: description
+          function: pattern
+          functionOptions:
+            match: Stoplight
+  - files:
+      - "**/*.json"
+    aliases:
+      Value:
+        - "$..value"
```
Comment (marbemac) [inline docs/guides/4-custom-rulesets.md:442]: Might be worth adding a note re override behavior. E.g. - if an alias called `Stoplight` were defined here, would it take precedence over the `Stoplight` alias defined in the root?

File: packages/core/src/ruleset/meta/shared.json [hunk 2]
```diff
@@ -50,9 +50,26 @@
     },
     "PathExpression": {
       "$id": "path-expression",
-      "type": "string",
-      "pattern": "^[$#]",
-      "errorMessage": "must be a valid JSON Path expression or a reference to the existing Alias optionally paired with a JSON Path expression subset"
+      "if": {
+        "type": "string"
+      },
+      "then": {
```
Comment (marbemac) [inline packages/core/src/ruleset/meta/shared.json:56]: This conditional logic stuff in json schema is cool and terrifying at the same time.

**Review by marbemac — COMMENTED** [review]
Let's carve out 20 min to walk through this one on a call, I'd like to understand a few bits better.

**Comment by chohmann** [comment]
I don't know much about Spectral... not sure if I'm the best person to review this or not.

**Comment by P0lip** [comment]
No worries, you can totally skip it.
I believe you were added because you're a member of`stoplight/documentation` and this group is now a code owner of the `docs/` directory.

**Comment by chohmann** [comment]
ahhh ok that makes sense. Thanks @P0lip !

**Comment by mmiask** [comment]
I'm having trouble understanding the changes. @P0lip  - what could speed up the review is some kind of changes summary, describing the main change and how other files relate to that. Would you mind putting such summary in PR description? Or if it is better for you, discuss it on a call and do live PR review?

**Review by mmiask — APPROVED** [review]
Read through this PR with @P0lip on a call. Awesome work, looks great 👍

**Pull request merged**